pgfinder/gui/internal.py

```diff
@@ -11,28 +11,23 @@
 
 
 def theo_masses_upload_reader(upload: dict) -> pd.DataFrame:
-    filename = upload["name"]
-    file_contents = open(MASS_LIB_DIR / filename, "rb").read() if upload["content"] is None else upload["content"]
+    # Attempt to load a built-in library if no content was uploaded
+    if upload["content"] is None:
+        upload["content"] = open(MASS_LIB_DIR / upload["name"], "rb").read()
 
-    theo_masses_df = pd.read_csv(io.BytesIO(file_contents))
-    theo_masses_df.columns = ["Inferred structure", "Theo (Da)"]
-    theo_masses_df.attrs["file"] = PurePath(filename).name
-    return theo_masses_df
+    with uploaded_file(upload) as file:
+        return pgio.theo_masses_reader(file)
 
 
 def ms_upload_reader(upload: dict) -> pd.DataFrame:
-    filename = upload["name"]
-    file_contents = upload["content"]
-    file_temp = tempfile.NamedTemporaryFile(delete=False)
-    file_temp.write(file_contents)
-    file = file_temp.name
-
-    if not filename.find("ftrs") == -1:
-        return_df = pgio.ftrs_reader(file)
-    elif not filename.find("txt") == -1:
-        return_df = pgio.maxquant_file_reader(file)
-    else:
-        raise ValueError("Unknown file type.")
-
-    return_df.attrs["file"] = PurePath(filename).name
-    return return_df
+    with uploaded_file(upload) as file:
+        return pgio.ms_file_reader(file)
+
+
+@contextmanager
+def uploaded_file(upload: dict):
+    with tempfile.TemporaryDirectory() as tempdir:
+        file = Path(tempdir) / upload["name"]
+        with open(file, "wb") as f:
+            f.write(upload["content"])
+        yield file
```

Comment:
This made me feel a bit clever — it's probably good for someone else to give this a look to be sure it's not "too clever" and is still comprehensible...